Simplifying macros

elichai opened the pull request:
This is the first step in an attempt to simplify the macros we use in this crate.
I tried removing anything that can be derived. and removing the general functions implementation that might made sense on crates like `rust-secp256k1` for ffi reasons but less so here.

cc #352 
I'm trying to "Modernize" some of the great code @apoelstra wrote way before any of us ever heard of rust :)
The next stage is when these are simple enough I can write a custom derive crate for the rest of the things.

Sadly a thing that can greatly simplify the Index operations was is only stable from 1.28 https://github.com/rust-lang/rust/issues/35729


The breaking changes are:
1. removing a bunch of `as_pt` etc. implementations from `ChainCode` and `Fingerprint` (as I said before I don't think they make any sense here).
2. removing a `#[repr(C)]` from the Uints.

anything else here shouldn't be considered a breaking change (exact implementations of traits like `Hash` etc aren't supposed to be stable anyway)

elichai commented on src/util/bip32.rs:
```diff
@@ -31,12 +31,14 @@ use util::base58;
 use util::key::{PublicKey, PrivateKey};
 
 /// A chain code
+#[derive(Copy, Clone, PartialEq, Eq)]
```
I didn't derive PartialOrd, Ord, Hash. 
should I add them? (I don't see an obvious need, except if someone throws these things into a Hashmap)

dr-orlovsky replied:
I think they can be added when they will be needed

apoelstra replied:
I think we should add them now. It's nice to be able to key hashmaps on everything.

stevenroose replied:
Idem. I've bumped into several missing derives lately and it's a bit annoying.

elichai replied:
Done

apoelstra replied:
What about `Hash`?

dr-orlovsky commented on src/util/uint.rs:
```diff
@@ -26,7 +26,7 @@ use util::BitArray;
 macro_rules! construct_uint {
     ($name:ident, $n_words:expr) => (
         /// Little-endian large integer type
-        #[repr(C)]
```
Why you have removed `#[repr(C)]`? It may break some code relying on rust-bitcoin through FFI...

elichai replied:
because it makes no sense to me to use this struct in FFI.
it's a uint with 64bit limbs little endian. it's very specific to the implementation here.

dr-orlovsky replied:
Than we need to remember to do a major version bump after this PR merge into the master

elichai replied:
yep. see breaking changes no. 2 in the post :)

dr-orlovsky commented on src/util/psbt/serialize.rs:
```diff
@@ -74,7 +74,7 @@ impl Serialize for (Fingerprint, DerivationPath) {
         let mut rv: Vec<u8> = Vec::with_capacity(4 + 4 * (self.1).as_ref().len());
 
-        rv.append(&mut self.0.to_bytes().to_vec());
+        rv.append(&mut (&self.0[..]).to_vec());
```
Should we add `to_vec` directly to the Fingerprint type?

stevenroose commented on src/internal_macros.rs:
```diff
@@ -45,41 +45,6 @@ macro_rules! impl_consensus_encoding {
-
-            #[inline]
-            /// Returns the underlying bytes.
-            pub fn into_bytes(self) -> [$ty; $len] { self.0 }
```
I don't seem to see where these last 5 methods have gone. Especially the last 3 are really useful.

elichai replied:
Really? because you basically get them for free with the `Index` impl.
but if you think they're useful i'll keep them

stevenroose replied:
Hmm, true that you get them with `Index`. The only one would be `into_inner`, not sure in what cases one wants to keep the inner value, but I think we ought support it, no?

stevenroose replied:
Did you add the into_ method? I think `into_inner` is a useful method always. Any newtype ever probably should provide it without any overhead like reallocation.

elichai replied:
Reverted that change

stevenroose commented:
Ok so sometimes converting *into* and *from* the inner object is useful. F.e. when interfacing with code that doesn't know any bitcoin conventions. I also did it before with FFI's. Apart from that, I'd ACK.

Aaand it would be good to have in 0.22 perhaps since it's quite a big breaking change. But not a big deal if it's lateron.

dr-orlovsky commented:
Would really love to see it merged. Can I help with adding the discussed changes to the PR?

elichai commented:
I kinda left it for a while to get feedback on if people even want that (especially because of the complications in the follow up PR with the derive macros)

dr-orlovsky commented:
I think that the level of current macro abuse in rust-bitcoin is too high and any more idiomatic code should be welcomed

stevenroose commented:
I feel a bit double with all this. I think it'd be good to have 2 PRs: one that simply removes all impls that can be derived. Because that PR is a no-brainer.

But removing methods and replacing them with Rust traits is more tricky. I remember as a new user that I often lacked "simple" or "basic" methods. And using traits, while nifty and cool, is less intuitive. A `Fingerprint` might be nothing more than 4 bytes, it's still a `Fingerprint` (instance) in Rust. It's not trivial/intuitive for users that you can just use all the byte traits on it. I think the overhead of providing basic methods like `as_bytes(&self) -> &[u8]`, `to_bytes(&self) -> Vec<u8>` and `into_inner(&self) -> [u8; N]` might be worth it. I also think it makes code more readable. `fp.as_bytes()` vs `&fp[..]` f.e.

elichai commented:
FYI CI failures are because of https://github.com/rust-bitcoin/rust-bitcoin/pull/398

dr-orlovsky commented on src/util/uint.rs:
```diff
@@ -91,6 +91,27 @@ macro_rules! construct_uint {
             }
         }
 
+        impl PartialOrd for $name {
```
Can we use `::std::cmp` for consistency here and in the next impl?

stevenroose replied:
We could, but it's not an exported macro, so it's not very crucial.

elichai replied:
Well if you didn't notice that's a copy-paste impl from https://github.com/rust-bitcoin/rust-bitcoin/pull/356/files#diff-ca925ea633ee82c8b9dfd44534146107L114
anyhow as Steven said these are internal macros so we(aka Andrew ;) ) never cared much about explicit here because they're just use in this crate, and all these traits are included the in the prelude (https://doc.rust-lang.org/std/prelude/index.html)

dr-orlovsky replied:
I have to export these macros in my develop branch of the rust-bitcoin (b/c I use them in libs depending on rust-bitcoin), and we may decide to make them exportable in the future.

apoelstra replied:
I really don't think we should export these. It is not the job of rust-bitcoin to be a general-purpose bignum (or whatever) library.

dr-orlovsky replied:
Sure, I'm not asking for that. I just do export in my own fork of rust-bitcoin :) But anyway it would be good to stick to the same code style within the macro itself, even if it is internal

stevenroose commented:
Needs rebase.

elichai commented:
I honestly don't know how, but somehow the changes I reverted before(https://github.com/rust-bitcoin/rust-bitcoin/pull/356#discussion_r353968030) came back after the last rebase, I might've had an old version locally, although I'm not quite sure how.

Anyhow I modified it to look back as the original, and you can check yourself, the hash after the first rebase is: c9be13b0935fa3f4b2f570119000b77015dafaa3 and the current hash is: 77589d2e122f63768fe7fde880078e439875842d so:
`git diff c9be13b0935fa3f4b2f570119000b77015dafaa3 77589d2e122f63768fe7fde880078e439875842d` and look at the relevant files

apoelstra commented:
I think I'm looking at the latest code and there are still missing `Hash` derives for the bip32 types.

stevenroose commented on src/util/bip32.rs:
```diff
@@ -30,12 +30,14 @@ use util::{base58, endian};
 impl_bytes_newtype!(ChainCode, 32);
 
 /// A fingerprint
+#[derive(Copy, Clone, PartialEq, Eq)]
```
Like Andrew points out, Hash is missing here and on `ChainCode`.

elichai commented:
Added the missing impls.

stevenroose reviewed (dismissed):
utACK ef5e244

elichai commented:
rebased

dr-orlovsky commented:
utACK https://github.com/rust-bitcoin/rust-bitcoin/pull/356/commits/02ddb270f46c3ca9e6d4f17bd446815e518a9863

apoelstra commented:
```
git rebase -x 'cargo test && cargo test --features "use-serde rand" && cargo +1.29.0 test' master
```
fails (first commit does not compile)
